compiler/rustc_mir_dataflow/src/value_analysis.rs

```diff
@@ -999,20 +999,33 @@ pub fn iter_fields<'tcx>(
                 f(None, field.into(), ty);
             }
         }
-        ty::Adt(def, args) => {
-            if def.is_union() {
-                return;
-            }
-            for (v_index, v_def) in def.variants().iter_enumerated() {
-                let variant = if def.is_struct() { None } else { Some(v_index) };
-                for (f_index, f_def) in v_def.fields.iter().enumerate() {
-                    let field_ty = f_def.ty(tcx, args);
-                    let field_ty = tcx
-                        .try_normalize_erasing_regions(param_env, field_ty)
-                        .unwrap_or_else(|_| tcx.erase_regions(field_ty));
-                    f(variant, f_index.into(), field_ty);
+        &ty::Adt(def, args) => {
+            fn iter_adt_fields<'tcx>(
+                def: ty::AdtDef<'tcx>,
+                args: ty::GenericArgsRef<'tcx>,
+                tcx: TyCtxt<'tcx>,
+                param_env: ty::ParamEnv<'tcx>,
+                f: &mut impl FnMut(Option<VariantIdx>, FieldIdx, Ty<'tcx>),
+            ) {
+                if def.is_union() {
+                    return;
+                }
+                for (v_index, v_def) in def.variants().iter_enumerated() {
+                    let variant = if def.is_struct() { None } else { Some(v_index) };
+                    for (f_index, f_def) in v_def.fields.iter().enumerate() {
+                        if let Some((def, args)) = f_def.inner_adt(tcx, args) {
+                            iter_adt_fields(def, args, tcx, param_env, &mut *f);
+                        } else {
+                            let field_ty = f_def.ty(tcx, args);
+                            let field_ty = tcx
+                                .try_normalize_erasing_regions(param_env, field_ty)
+                                .unwrap_or_else(|_| tcx.erase_regions(field_ty));
+                            f(variant, f_index.into(), field_ty);
+                        }
+                    }
                 }
             }
+            iter_adt_fields(def, args, tcx, param_env, &mut f);
```

Comment:
Before changing this: what is the MIR produced by accessing an field of anonymous struct/union? Could you add a mir building test?

Reply:
Nice suggestion! I was thinking about how to test those field-accessing behaviors and then came your advice

Reply:
By the MIR building test, accessing a field of an anonymous struct/union produces nested field accesses.